keps/sig-node/5246-cgroup-cpu-share-to-weight-conversion/README.md

```diff
@@ -0,0 +1,939 @@
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+As [suggested](https://github.com/kubernetes/kubernetes/issues/131216#issuecomment-2806442083)
+by the [original KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2254-cgroup-v2) author,
+we should migrate to systemd's formula for converting CPU shares to CPU weight.
+
+The current formula that we're using is:
+`cpu.weight = (1 + ((cpu.shares - 2) * 9999) / 262142)`
+
+While systemd's formula is:
+`cpu.weight = 1 + ((cpu.shares – 2) × 99) / (1024 – 2)`
+
+The main difference between the formulas is that the first one focuses on converting the values from one range to another,
+keeping the values in the same percentile in this range.
+The second one focuses on aligning the values to the default CPU weight of `100` for a container requesting 1 CPU.
+Accordingly, 512m CPU will result in a value of `50`, which aligns with the default v1 CPU shares of `512` for a container requesting 1 CPU.
+
+Let cpu.shares be 1024. Therefore: `1 + ((cpu.shares – 2) × 99) / (1024 – 2) == 100`.
+
+Let cpu.shares be 512. Therefore: `1 + ((cpu.shares – 2) × 99) / (1024 – 2) == 50.4031 ~= 50`.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+As a Kubernetes user, when I used to work with v1, a container asking for 1 CPU had the same CPU shares as a non-kubernetes
+workload. After moving to cgroup v2 I expect this behavior to stay the same, but instead,
+the CPU priority for Kubernetes workloads is much lower than non-kubernetes workloads compared to v1.
+
+#### Story 2
+As a Kubernetes user, I want to be able to configure sub-cgroups inside a container to further
+distribute resources inside the container. While I could do that nicely with v1, with v2 the granularity is not fine-grained enough.
+As an example, `100m` CPU on v2 results with `4` CPU weight, while on v1 it would be `102` shares.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- A significant amount of the work would need to land in other layers, mainly OCI runtimes and the CRI.
+- We'll probably need a CRI configuration to ensure coordination between the CRI and the OCI runtimes implementations,
+and to ensure it lands at the same version, as suggested
```

Comment:
Should we expand this in the design section to understand what needs to be changed and how the rollout will happen?

Reply:
Hey @yujuhong!

What I have in mind is:
1. Change implementation in OCIs.
2. Expose a configuration the CRI can use.
3. Ensure CRI supports the new OCI configuration.
4. Test k8s with the new CRI configuration.
5. In the same k8s release:
	1. Change the k8s code related to this formula, mainly pod-resources resize etc.
	2. Ensure CRIs enable the OCI config.

Does that sound right? WDYT?